### PR TITLE
Update macos build

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR reworks the build settings, primarily on macos.

On macos, -fPIC and -PIE are not necessary: they do nothing. This gives us the opportunity to use the compile static and shared `.o` with the same flags and reuse them in 'shared' and 'static' builds... and that's exactly what this PR does.

On Linux, we keep (for now) fPIC, fPIE and separate directories as before.

### CODE REVIEW

File: cpp/src/Glacier2CryptPermissionsVerifier/Makefile.mk [hunk 1]
```diff
@@ -6,7 +6,7 @@ $(project)_libraries                            += Glacier2CryptPermissionsVerif
 
 Glacier2CryptPermissionsVerifier_targetdir        := $(libdir)
 Glacier2CryptPermissionsVerifier_dependencies     := Glacier2 Ice
-Glacier2CryptPermissionsVerifier[shared]_cppflags := -DCRYPT_PERMISSIONS_VERIFIER_API_EXPORTS
+Glacier2CryptPermissionsVerifier_cppflags         := -DCRYPT_PERMISSIONS_VERIFIER_API_EXPORTS
```
Comment (bernardnormier) [inline cpp/src/Glacier2CryptPermissionsVerifier/Makefile.mk:9]: Reverted an earlier commit that add "[shared]". We want to use exactly the same flags for static and shared builds on macos.

File: cpp/src/Ice/Exception.cpp [hunk 2]
```diff
@@ -47,7 +47,7 @@
 #        endif
 #    endif
 
-#    if !defined(__FreeBSD__) && defined(ICE_API_EXPORTS)
```
Comment (bernardnormier) [inline cpp/src/Ice/Exception.cpp:50]: Not sure if we can remove this `defined(ICE_API_EXPORTS)`. It seems to work fine on macos. Would be good to test on Linux with a static build.

File: cpp/src/Ice/Makefile.mk [hunk 3]
```diff
@@ -16,7 +15,11 @@ Ice_extra_sources       := $(filter-out src/Ice/SSL/OpenSSL%.cpp src/Ice/SSL/Sch
 else
 Ice_extra_sources       := $(filter-out src/Ice/SSL/SecureTransport%.cpp src/Ice/SSL/Schannel%.cpp, $(wildcard src/Ice/SSL/*.cpp))
 endif
+
 Ice_excludes            = src/Ice/DLLMain.cpp
+Ice[shared]_excludes    = src/Ice/RegisterPluginsInit_min.cpp
```
Comment (bernardnormier) [inline cpp/src/Ice/Makefile.mk:20]: That's how I can use the same flags for static/shared build and yet get a small configuration-specific behavior: include a different impl file depending on the configuration.

File: cpp/src/Ice/RegisterPluginsInit_all.cpp [hunk 4]
```diff
@@ -20,11 +19,7 @@ IceInternal::RegisterPluginsInit::RegisterPluginsInit()
     Ice::registerPluginFactory("IceTCP", createIceTCP, true);
     Ice::registerPluginFactory("IceSSL", createIceSSL, true);
 
-    //
-    // Include the UDP and WS transport plugins with non-static builds.
-    //
-#if defined(ICE_API_EXPORTS)
+    // Include the UDP and WS transport plugins with "shared" builds.
```
Comment (bernardnormier) [inline cpp/src/Ice/RegisterPluginsInit_all.cpp:22]: That's the implementation (_all) we want to use/include in all builds, except the configuration=static build.

File: scripts/Component.py [hunk 5]
```diff
@@ -98,6 +98,7 @@ def getFilters(self, mapping, config):
                     "IceSSL/configuration",
                     "IceDiscovery/simple",
                     "IceGrid/simple",
+                    "IceUtil/stacktrace",
```
Comment (bernardnormier) [inline scripts/Component.py:101]: Enable this test for static builds. Otherwise, it's hard to see if we have an issue with backtraces and static builds!

**Review by pepone — APPROVED** [review]

**Review by externl — APPROVED** [review]

**Pull request merged**